bpo-28009: Fix uuid.uuid1() and uuid.get_node() on AIX

aixtools opened the pull request:
FYI: linux et al use b':' to separate the 6 fields of a MAC address, also the fields are always two characters. On AIX the character b'.' is used, and when the decimal value is < 16, only one character is used.

Was: b'00:0a:10:20:30:40', on AIX - could be: b'0.a.10.20.30.40'

p.s. I hope this can be considered for backporting.

<!-- issue-number: [bpo-28009](https://www.bugs.python.org/issue28009) -->
https://bugs.python.org/issue28009
<!-- /issue-number -->

aixtools committed:
uuid: correct the field seperator value of ethernet adapter MAC address on AIX

aixtools committed:
Merge branch 'master' into bpo-28009-1
correct the field seperator character of ethernet adapter MAC address on AIX

aixtools committed:
correct test logic, cut/paste error

aixtools committed:
specify shorter getters list per https://bugs.python.org/issue28009#msg306562
also, calculate correct value when the fields are not all two characters

vstinner commented on Lib/test/test_uuid.py:
```diff
@@ -6,11 +6,13 @@
 import subprocess
+import sys
+
+_notAIX = not sys.platform.startswith("aix")
```
I dislike negative constant: I suggest to rename it to AIX and use it with"not AIX". No need to make the constant private in an unit test.

aixtools replied:
done.

vstinner commented on Lib/test/test_uuid.py:
```diff
@@ -549,7 +558,10 @@ def check_node(self, node, requires=None):
+        if _notAIX:
+            node = self.uuid._ifconfig_getnode()
+        else:
+            self.skipTest('because AIX "ifconfig" does not provide macaddr')
```
You should use @unittest.skipIf(AIX, "...").

aixtools replied:
much nicer. done

vstinner commented on Lib/test/test_uuid.py:
```diff
@@ -559,7 +571,10 @@ def test_ip_getnode(self):
 
     @unittest.skipUnless(os.name == 'posix', 'requires Posix')
```
You should use @unittest.skipIf(AIX, "...").

aixtools replied:
done

vstinner commented on Lib/uuid.py:
```diff
@@ -52,6 +52,9 @@
 
 __author__ = 'Ka-Ping Yee <ping@zesty.ca>'
 
+_notAIX = not sys.platform.startswith("aix")
```
Please rename the constant to _AIX and use "not _AIX".

aixtools replied:
done.

vstinner commented on Lib/uuid.py:
```diff
@@ -52,6 +52,9 @@
 __author__ = 'Ka-Ping Yee <ping@zesty.ca>'
 
+_notAIX = not sys.platform.startswith("aix")
+_mac_delim =  b':' if _notAIX else b'.'
```
Rename the constant to _MAC_DELIM.

aixtools replied:
done

vstinner commented on Lib/uuid.py:
```diff
@@ -455,8 +458,15 @@ def _netstat_getnode():
+                    if word.count(_mac_delim) == 5:
+                        if len(word) == 17:
+                            mac = int(word.replace(_mac_delim, b''), 16)
+                        elif len(word) < 17 and len(word) >= 11:
```
I don't understand this change. Is it only needed on AIX? I would prefer to keep ==17 test on other platforms.

aixtools replied:
As I do not have access to other platforms I was trying to write something that was generic. 
Roughly - first test for the _MAC_DELIM count.
Then - if the overall length is 17 the characters will be the correct hex value - when evaluated, so the splitting into individual hex values and adding and shifting is not needed.
Else: each individual hex value must be added in separately AFTER an 8/bit left shift of the previous value.

Otherwise a macaddr from:
en0   1500  link#2      fa.d1.8c.f7.62.4 1601514231     0 664547436     0     0

returns as:
fad18cf7624 (fad 18cf 7624, dec ‭17236120008228‬) rather than as
fad18cf76204 (fad1 8cf7 6204 , dec 275777920131588)

If you still want it changed, please comment again after push in about 5 minutes)

vstinner commented on Misc/NEWS.d/next/Library/2018-08-04-12-26-11.bpo-28009.4JcHZb.rst:
```diff
@@ -0,0 +1 @@
+uuid: correct the field seperator (_mac_field) value of an ethernet adapter MAC address for AIX
```
This NEWS entry is misleading: your change includes other changes. Maybe you could just say "Fix uuid.uuid1() and uuid.get_node() on AIX"?

aixtools replied:
done

aixtools committed:
write _notAIX as non-negative constant and use 'not AIX'

aixtools committed:
simplify NEWS entry

aixtools committed:
skip a test the right way

aixtools committed:
capitalize constants

aixtools committed:
oops - restore missing assignment

aixtools committed:
skip both tests the right way

aixtools committed:
write _notAIX as non-negative constant and use 'not _AIX'

aixtools committed:
added additional 'skip' entries for commands that do not exist on AIX

aixtools reviewed (commented):
changes done, or comments added

vstinner commented on Lib/uuid.py:
```diff
@@ -663,6 +676,8 @@ def _random_getnode():
 _NODE_GETTERS_UNIX = [_unix_getnode, _ifconfig_getnode, _ip_getnode,
                       _arp_getnode, _lanscan_getnode, _netstat_getnode]
 
+_NODE_GETTERS_AIX = [_unix_getnode, _netstat_getnode]
```
I would prefer to see a single _NODE_GETTERS constant and move the the platform test ("if sys.platform == 'win32':") here: move it from getnode().

vstinner commented on Lib/uuid.py:
```diff
@@ -455,8 +458,18 @@ def _netstat_getnode():
+                        # the extracted hex string will not be a 12 hex digit
+                        # string, so extract the fields and add them in
+                        # piece by piece
+                        elif len(word) < 17 and len(word) >= 11:
```
Please revert this change.

If you want to keep it, please make it specific to AIX and write an unit test for it.

aixtools replied:
made specific for AIX

aixtools committed:
Use and set a single _NODE_GETTERS constant with platform test outside getnode()
Move _netstat_getnode() find alg so that it can be tested using unittest.mock
Add new test `test_uuid/find_mac_netstat` for AIX and netstat

vstinner commented on Lib/uuid.py:
```diff
@@ -388,6 +392,48 @@ def _find_mac(command, args, hw_identifiers, get_index):
         pass
     return first_local_mac or None
 
+def _find_mac_netstat(command, args, hw_identifiers, get_index):
```
I don't understand why this function is needed? Why not leaving the code in _netstat_getnode()? I don't understand the purpose of get_index. It seems like your design is to be able to write more easily tests. But you can mock a lock of things, especially the uuid._AIX constant. You can test the AIX-specific code on all platforms instead of skipping the test on other platforms. Otherwise, if the AIX code is broken, nobody will notice it until someone runs the tests on AIX.

aixtools replied:
a) not to be creative - more because you asked for an AIX specific test but also to duplicate/copy the way _find_mac() is used by the other routines - as well as tested by test_uuid.
b) _find_mac is targeting output that has "keyword value" pairs on the same line, while _find_mac_netstat (maybe a different name is needed) is for output where the first line gives the position of values to consider. In short, _find_mac cannot parse 'netstat' output, so rather than have the code twice I thought it prudent to move the logic outside of _netstat_getnode().
c) as to being able to mock in other ways - no doubt. But this is what I could come up with the skill level I have. Suggestions for a better way are welcome. I like to learn - and even doing this helped my learn a lot.
d) the code is not checked via the fake data (if not AIX), but it is checked via the other test(s) that call _netstat_getnode(). However, if additional fake data is provided for other platforms the the test can also be used for them. This is what I can provide - and test manually.

ncoghlan replied:
I think a good way to help make the code clearer here will be to do the following:

1. Change the name of the new function to `_find_mac_in_table`, drop the `get_index` parameter, and rename `hw_identifiers` to `mac_columns`.
2. Give the new function a docstring that says "Find MAC addresses in tabular command output using the given column name"
3. In the existing `_find_mac` function change the name of the `get_index` parameter to `get_value_index` and the name of the `hw_identifiers` parameter to `mac_keywords`
4. Give the existing function a docstring that says "Find MAC addresses in command output by locating values relative to particular text keywords"

A future patch might further refactor these functions to avoid duplicating the command execution logic (and perhaps modernise them to use `enumerate` where appropriate), but I don't think that's necessary at this stage.

vstinner commented on Lib/test/test_uuid.py:
```diff
@@ -514,14 +515,39 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
+                    command='netstat',
+                    args='-ia',
+                    hw_identifiers=b'Address',
+                    get_index=lambda x: x * 1,
```
I don't understand why you need to use i+1 here. It looks like a bug.

aixtools replied:
Guess my replies did not make it here.
a) It is "* 1" not plus 1
b) it may indicate a bug (not in the test I fear, but elsewhere in core) - when I was trying to use the find_mac code to parse netstat output I first tried "get_index=lambda x: x", but it always evaluated as get_index=lambda x: x + 1. I may have also tried x + 0, but I do not recall. Before merging I can experiment with other values - and if they are not working that is a different bug - not the one I am trying to resolve.

aixtools committed:
modify the lambda function so it does not look like a bug

aixtools committed:
resolve differences in imports

aixtools committed:
Switch back to using sys.platform.startswith() to resolve conflicts

aixtools committed:
still trying to get differencs to resolve...

taleinat commented on Lib/test/test_uuid.py:
```diff
@@ -514,14 +514,39 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
+        popen = unittest.mock.MagicMock()
+        popen.stdout = io.BytesIO(data.encode())
+
+        with unittest.mock.patch.object(shutil, 'which',
```
Is this just an optimization to avoid calling `which` here, since `netstat` won't actually be invoked?

Whether I've understood correctly or not, please add a comment explaining this, since it is non-obvious.

aixtools replied:
Actually, it is copycat. If you still feel it need an additional explanation. Just repopen the comment. I am hoping this resolves this comment.

(And thanks for the polish hints!)

Below there are the lines:
        with unittest.mock.patch.object(shutil, 'which',
                                        return_value='/sbin/ifconfig'):

taleinat replied:
Oh, in that case, do keep the consistency with that.

taleinat commented on Lib/test/test_uuid.py:
```diff
@@ -514,14 +514,39 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
+        self.assertEqual(mac, 0xfead0c012304)
+
     @unittest.skipUnless(os.name == 'posix', 'requires Posix')
+    @unittest.skipIf(AIX, 'because AIX "ifconfig" does not provide macaddr')
```
Drop the word "because" in the description. (also in other similar cases here)

aixtools replied:
done

taleinat commented on Lib/test/test_uuid.py:
```diff
@@ -514,14 +514,39 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
     def test_find_mac(self):
-        data = '''
-fake hwaddr
+        data = '''fake      Link encap:UNSPEC  hwaddr 00-00
```
Is changing this line relevant to fixing things on AIX? It seems like it couldn't be, so why is it here?

aixtools replied:
no. the change is not needed here. The reason was for symmetry with line 519 - where it is important to not have an empty first line (the data starts as "\nfake...," while the AIX block requires that the data starts as "Name  Mtu...", i.e., without the \n prefix.

Reverted back.

Please feel free to ask me to change it back if you concur that a hidden \n is not self-explanatory.

taleinat replied:
You can use the following if the beginning `\n` is problematic:
```python
        data = '''\
fake hwaddr
...
        '''
```

The `\` at the end of the line avoids adding a newline.

taleinat commented on Lib/uuid.py:
```diff
@@ -388,6 +392,48 @@ def _find_mac(command, args, hw_identifiers, get_index):
         pass
     return first_local_mac or None
 
+def _find_mac_netstat(command, args, hw_identifiers, get_index):
```
At least in this PR's current form, it seems that the values given for the two new parameters `hw_identifiers` and `get_index` are always the same (`b'Address'` and `lambda x: x` respectively; I'm considering `lambda x: x + 0` equivalent to `lambda x: x` since the inputs are always integers.)

Now that this is working, please try to "unwind" some of the changes made while getting to this point, to see if things work properly without them.

Start with replacing `lambda x: x + 0` with `lambda x: x`, which I would be surprised if had any effect.

Then try moving this logic back into `_netstat_getnode()`, since apparently the additional parameterization is no longer needed.

aixtools replied:
I had had issues, initially using lambda i: i*1, as lambda i: i did not work first times around. Testing..

aixtools replied:
testing is working on AIX. Pushing for CI

taleinat replied:
Are you still working on this? My original comment here is still relevant.

taleinat replied:
I've read this more thoroughly and now understand that this is modified copy/paste of the existing `_find_mac()` function. This should be avoided if reasonably possible.

If the only significant change is the mac address parsing due to different formatting, perhaps we could just add a mac address parser as a parameter to `_find_mac()`? E.g.:

```python
def _find_mac(command, args, hw_identifiers, get_index,
              parse_mac=lambda word: int(word.replace(b':', b''), 16)):
...
```

aixtools replied:
Finally - found where the email came from. ;) I guess this is part of the mentor process - find the comments.

This is "double" as I did reply to the email - but I'll add my response here for clarity.
-
Yes, the routine is based on _find_mac(). But _find_mac is designed for data that has the keyword and the value on the same line. Programs such as netstat have a different layout. And yes, AIX also has different formatting - using the character '.' to divide the fields and also a variable length (1.2.3.4.5.6 is a valid format on AIX (and would be evaluated as 0x010203040506)

The (linux (what I found)) ifconfig commands,as well as the mock-data have the keyword and
the value on a single line while netstat (on both AIX and macos) have the keyword on the "header" line
and in that "position" the keyword on a later line. When there are fewer words, even if it looks as if it is
that column, and/or is not formatted properly (not six 'words' separated

Snippet from macos:

```
Name  Mtu   Network       Address            Ipkts Ierrs    Opkts Oerrs  Coll
lo0   16384 <Link#1>                         56907     0    56907     0     0
lo0   16384 127           localhost          56907     -    56907     -     -
...
lo0   16384 127.94.0.1/32 127.94.0.1         56907     -    56907     -     -
                        224.0.0.251
                        all-systems.mcast.net
gif0* 1280  <Link#2>                             0     0        0     0     0
stf0* 1280  <Link#3>                             0     0        0     0     0
XHC20 0     <Link#4>                             0     0        0     0     0
en2   1500  <Link#5>    1a:00:01:0d:82:10        0     0        0     0     0
en3   1500  <Link#6>    1a:00:01:0d:82:11        0     0        0     0     0
en0   1500  <Link#7>    38:c9:86:2b:38:f5  3816938     0  3025627     0     0

```
If called with this data the value 0x1a00010d8210 would be used (IF it _is_universal(0x1a00010d8210) OR not _is_universal(0x38c9862b38f5)).

Snippet from AIX:
```
Name  Mtu   Network     Address           Ipkts Ierrs    Opkts Oerrs  Coll
en0   1500  link#2      fa.d1.8c.f7.62.4 129294604     0 57078842     0     0
```
Here the value returned is 0xfad18cf76204 (not 0xfad18cf7624)
by a character).

Further, I have always tried to say away from something that could be
construed as a feature change. From my perspective it is hard enough to
get a straight forward change considered. A feature change seems an
impossibility - e.g., immediate concerns about a potential backwards
compatibility issue. I leave changes of that caliber to others who have
been involved much longer and deeper than myself.

Hope this answers your question. In any case, I await your response before considering to "resolve conversation"

Sincerely,
Michael

ncoghlan replied:
`netstat -ia` doesn't actually report a MAC address on Linux:

```
[ncoghlan@localhost ~]$ netstat -ia
Kernel Interface table
Iface             MTU    RX-OK RX-ERR RX-DRP RX-OVR    TX-OK TX-ERR TX-DRP TX-OVR Flg
enp58s0f1        1500        0      0      0 0             0      0      0      0 BMU
lo              65536     1526      0      0 0          1526      0      0      0 LRU
wlp59s0          1500   400593      0      0 0        196557      0      0      0 BMRU
```

So I think the three cases you'll want to test for cross-platform maintainability are:

1. Mac OS X style, ensure the function gives the right answer
2. AIX style, ensure the function gives the right answer
3. Linux style, ensure the function correctly returns `None` (since there's no "Address" column)

The typical way we do that kind of testing elsewhere in the test suite is to iterate over a sequence of "subtest-identifier, input data, expected result" 3-tuples, and then use the [unittest subtest feature](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests) feature to report the results for each case.

Regarding the difficulty of getting straight forward changes considered: the coupling to operating system dependent hardware introspection means that nothing about the UUID module is straight forward in practice. Instead, it turns into a complex discussion of both test suite refactoring to help ensure as much code as possible can be tested in a platform independent way, while also making sure that the platform dependent parts are actually being exercised on the platforms that support them.

aixtools replied:
It was not a "copy" of the _find_mac routine. Instead, I moved the "_find_mac" logic that was in the _netstat_getnode() function - to make all of the "program" related functions more similiar (calling a find_mac_{inline|nextlines) based on whether the keyword (aka hw_indentifier) that used to determine the index that is passed to the "lanbda" function get_index().

Long sentence: in the "inline" version, the keyword and the index are in the same line; in "nextlines" the keyword only appears once in a "header" line.

Once I figured out that distinction - the code was clearer to me - and I hope now, to others as well.

taleinat commented on Lib/uuid.py:
```diff
@@ -439,32 +485,9 @@ def _lanscan_getnode():
-    return first_local_mac or None
+    # This works on AIX and might work on Tru64 UNIX.
+    mac = _find_mac_netstat('netstat', '-ia', b'Address', lambda i: i+0)
+    return mac if mac else None
```
This should be spelled: `return mac or None`.

aixtools replied:
:) Thx. Had to think on that a bit. Boolean!

aixtools replied:
Don't know where it went. I had corrected this. Maybe I undid that with testing.

taleinat reviewed (changes requested):
I like this fix! It needs a bit more polish bit it seems nearly ready to go.

aixtools committed:
changes per request. Thx for the review!

aixtools committed:
resolve merge conflict

aixtools committed:
resolve merge conflict, 2nd try

taleinat commented on Lib/test/test_uuid.py:
```diff
@@ -4,12 +4,17 @@
 import contextlib
 import io
 import os
+import pickle
```
Why is this import added?

aixtools replied:
this was to try and get merge to pass. I have now tried to add the current state of the test to my test (manual patch) as git rebase/merge of a single file is beyind my understanding of git.

aixtools replied:
well, current status - since merge won't accept it, and the manual merge of the pickle addition does not work I'll probably have to start from scratch again (about the 4th time now iirc).

I'll get back to this as soon as I can - but it seems I have been "pickled" :p@me.

taleinat commented on Lib/test/test_uuid.py:
```diff
@@ -4,12 +4,17 @@
 py_uuid = support.import_fresh_module('uuid', blocked=['_uuid'])
 c_uuid = support.import_fresh_module('uuid', fresh=['_uuid'])
-
+# would prefer using platform.system() but CI tests have a merge conflict with "import platform"
```
Merge/rebase this PR branch to overcome merge conflicts, and use the most appropriate code here.

aixtools replied:
git scares me :p@me,
nods I'll find a way.

taleinat replied:
Why do you prefer `platform.system()` over `sys.platform`? What value does `sys.platform` have on AIX? If it is reasonable ISTM that we should use `sys.platform` here.

In other words, the existing `AIX = sys.platform.startswith("aix")` seems fine to me, if it works well let's just remove the comments above it.

aixtools reviewed (commented):
Misunderstood your request. @taleinat 

I'll try testing it without the lambda statement. Understanding it was hard - but it seems on Linux the position it identifies needs to be increased by one to find the actual value in the following lines. With AIX the positions are equal.

Supposing the function can work without the lambda statement I would still want to consider leaving it in so that "lambda" is used consistently throughout the file.

During my initial debugging/learning I used lambda x: x, but this failed until I used lambda x: x * 1, or lambda x: x + 0.

In any case, I am not trying to be fancy - merely consistent with previous code.

After starting a new branch, not submitted yet, removing the lambda will make the code look something like this:

```
-def _find_mac_netstat(command, args, hw_identifiers, get_index):
+# def _find_mac_netstat(command, args, hw_identifiers, get_index):
+def _find_mac_netstat(command, args):
     first_local_mac = None
     mac = None
     try:
@@ -417,18 +418,21 @@ def _find_mac_netstat(command, args, hw_identifiers, get_index):
         if not proc:
             return None
         with proc:
-            words = proc.stdout.readline().rstrip().split()
-            try:
-                i = words.index(hw_identifiers)
-            except ValueError:
-                return None
+            # words = proc.stdout.readline().rstrip().split()
+            # try:
+                # i = words.index(hw_identifiers)
+                # i = 4
+            # except ValueError:
+                # return None
             for line in proc.stdout:
                 try:
                     words = line.rstrip().split()
-                    word = words[get_index(i)]
+                    # word = words[get_index(i)]
+                    word = words[4]
                     if len(word) == 17 and word.count(_MAC_DELIM) == 5:
                         mac = int(word.replace(_MAC_DELIM, b''), 16)
-                    elif _AIX and  word.count(_MAC_DELIM) == 5:
+                    # elif _AIX and  word.count(_MAC_DELIM) == 5:
+                    elif word.count(_MAC_DELIM) == 5:
                         # the extracted hex string is not a 12 hex digit
                         # string, so add the fields piece by piece
                         if len(word) < 17 and len(word) >= 11:
@@ -503,7 +507,8 @@ def _lanscan_getnode():
 def _netstat_getnode():
     """Get the hardware address on Unix by running netstat."""
     # This works on AIX and might work on Tru64 UNIX.
-    mac = _find_mac_netstat('netstat', '-ia', b'Address', lambda i: i)
+    # mac = _find_mac_netstat('netstat', '-ia', b'Address', lambda i: i)
+    mac = _find_mac_netstat('netstat', '-ia')
     return mac or None

 def _ipconfig_getnode():
```

And a simple change in test_uuid.py (to not use the arguments either) also fails:

```
--- a/Lib/test/test_uuid.py
+++ b/Lib/test/test_uuid.py
@@ -674,8 +674,6 @@ en0   1500  192.168.90  x071             1714807956     0 711348489     0     0
                 mac = self.uuid._find_mac_netstat(
                     command='netstat',
                     args='-ia',
-                    hw_identifiers=b'Address',
-                    get_index=lambda x: x,
                 )

         self.assertEqual(mac, 0xfead0c012304)

```
AND - the test's no longer work - even with this simple modification above (different errors without the change):
```
======================================================================
FAIL: test_find_mac_netstat (test.test_uuid.TestInternalsWithExtModule)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/prj/python/git/python3-3.8/Lib/test/test_uuid.py", line 679, in test_find_mac_netstat
    self.assertEqual(mac, 0xfead0c012304)
AssertionError: None != 280019184198404

======================================================================
FAIL: test_find_mac_netstat (test.test_uuid.TestInternalsWithoutExtModule)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/prj/python/git/python3-3.8/Lib/test/test_uuid.py", line 679, in test_find_mac_netstat
    self.assertEqual(mac, 0xfead0c012304)
AssertionError: None != 280019184198404

----------------------------------------------------------------------

```
Given a choice, I would much rather continue using the "familiar" keyword + lambda arguments.

taleinat committed:
remove unused added import

taleinat commented on Lib/test/test_uuid.py:
```diff
@@ -6,10 +6,14 @@
+# would prefer using platform.system() but CI tests have a merge conflict with "import platform"
+# removing it for the moment
+# AIX = platform.system() == 'AIX'
+AIX = sys.platform.startswith("aix")
```
Consider naming this `_AIX` for consistency.

aixtools committed:
Per requested changes

aixtools committed:
Merge branch 'bpo-28009-1' of github.com:aixtools/cpython into bpo-28009-1

ncoghlan committed:
Merge branch 'master' into bpo-28009-1

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -9,11 +9,10 @@
 py_uuid = support.import_fresh_module('uuid', blocked=['_uuid'])
 c_uuid = support.import_fresh_module('uuid', fresh=['_uuid'])
-
+_AIX = sys.platform.startswith("aix")
```
Rather than duplicating the platform detection logic from `uuid` in the tests, I think we'll get a better outcome by basing the test skips on the contents of `uuid._NODE_GETTERS`. Otherwise we're setting ourselves up for a situation where one of the node querying functions gets called on more platforms than it is actually being tested on, as well as the inverse, where we end up testing a function on a platform where it's never even going to be tried.

aixtools replied:
This was done in a PR focusing on the tests.

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -669,14 +667,41 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
 class BaseTestInternals:
     uuid = None
 
+    @unittest.skipUnless(_AIX, 'requires AIX')
```
Check for `uuid._netstat_getnode in uuid._NODE_GETTERS` in deciding whether or not to run this one.

aixtools replied:
Was done in a separate PR

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -669,14 +667,41 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
+        self.assertEqual(mac, 0xfead0c012304)
+
     @unittest.skipUnless(os.name == 'posix', 'requires Posix')
+    @unittest.skipIf(_AIX, 'AIX "ifconfig" does not provide macaddr')
```
Check for `unittest.skipUnless(uuid._ifconfig_getnode in uuid._NODE_GETTERS, "ifconfig is not used for introspection on this platform")` (replaces both decorators, not just the new one)

aixtools replied:
Done in separate PR

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -703,21 +728,25 @@ def check_node(self, node, requires=None):
                         "%s is not an RFC 4122 node ID" % hex)
 
     @unittest.skipUnless(os.name == 'posix', 'requires Posix')
+    @unittest.skipIf(_AIX, 'AIX "ifconfig" does not provide macaddr')
```
Same check for `_ifconfig_getnode` as above (replaces both decorators)

aixtools replied:
Done in separate PR

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -703,21 +728,25 @@ def check_node(self, node, requires=None):
         self.check_node(node, 'ifconfig')
 
     @unittest.skipUnless(os.name == 'posix', 'requires Posix')
+    @unittest.skipIf(_AIX, 'requires command "ip"')
```
`unittest.skipUnless(uuid._ip_getnode in uuid._NODE_GETTERS, "ip is not used for introspection on this platform")`

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -703,21 +728,25 @@ def check_node(self, node, requires=None):
         self.check_node(node, 'ip')
 
     @unittest.skipUnless(os.name == 'posix', 'requires Posix')
+    @unittest.skipIf(_AIX, 'AIX "arp" does not provide macaddr')
```
`unittest.skipUnless(uuid._arp_getnode in uuid._NODE_GETTERS, "arp is not used for introspection on this platform")`

aixtools replied:
Done in separate PR

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -703,21 +728,25 @@ def check_node(self, node, requires=None):
         self.check_node(node, 'arp')
 
     @unittest.skipUnless(os.name == 'posix', 'requires Posix')
+    @unittest.skipIf(_AIX, 'requires command "lanscan"')
```
`unittest.skipUnless(uuid._lanscan_getnode in uuid._NODE_GETTERS, "lanscan is not used for introspection on this platform")`

ncoghlan reviewed (changes requested):
Hi @aixtools, thanks for your patience as we work through this more-complex-than-it-may-appear-on-the-surface change. The `uuid` module has been around for a long time, so some of the ways it is factored internally aren't necessarily the best, which can make reviewing even seemingly simple changes more difficult than we might expect.

In this case, I've proposed making some changes to the existing `_find_mac` function to help make the way it works clearer to readers, and hence also make it more obvious how it differs from the new `_find_mac_in_table` helper function. I also think we should take advantage of your new module level `uuid._NODE_GETTERS` attribute to help make it easier to keep the test suite in sync with the platform introspection logic in the module itself.

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -669,14 +667,41 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
+                    get_index=lambda x: x,
+                )
+
+        self.assertEqual(mac, 0xfead0c012304)
```
Assuming you go down the path of defining multiple subtests as described in 
https://github.com/python/cpython/pull/8672/files/70a45f01049079ddf00d3ba73f0fe3cf648d1985#r243934748, the expected results will need to be amended based on the value of `uuid._MAC_DELIM`: the AIX delimiter shouldn't report any results given the Mac OS X style `netstat` output, and vice-versa.

aixtools replied:
I have re-done this so many times - I'll have to take some time to review your request. I'll act on it asap, but it will probably be days to weeks.

Many thanks for the review - Best wishes for 2019!

aixtools replied:
Done in separate PR

aixtools committed:
sync with master

aixtools committed:
rename find_mac_routines (in test_uuid.py)

aixtools committed:
fix sync issues

aixtools committed:
sync with master

aixtools committed:
add inline comments, correct typos

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -674,26 +674,57 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
 
+    # data is specific to AIX - with '.' as _MAC_DELIM
+    # and strings shorter than 17 bytes (no leading 0)
+    @unittest.skipUnless(_AIX, 'requires AIX')
```
Similar to the skip logic changes that you extracted out to other PRs, a better check here will be `@unittest.skipUnless(self.uuid._MAC_DELIM == b'.')`, since that's what the sample data is assuming.

ncoghlan replied:
That also suggests abstracting out a marker attribute for `uuid._MAC_OMITS_LEADING_ZEROES`, so the full check would be:

    @unittest.skipUnless(self.uuid._MAC_OMITS_LEADINGS_ZEROES)
    @unittest.skipUnless(self.uuid._MAC_DELIM == b'.')

aixtools replied:
You (or I) resolved this. Both marker attributes are defined, and rather than skip, are assigned to the mock tests.

If I have missed something - I'll get on it, if not - please consider approving the changes, Thx.

ncoghlan replied:
The issue is that this is a valid test anywhere that those two conditions are true, so rather than testing for "AIX_" here, you can just test for the MAC notation conventions that the test cares about.

ncoghlan commented on Lib/test/test_uuid.py:
```diff
@@ -674,26 +674,57 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
+
     @unittest.skipUnless(os.name == 'posix', 'requires Posix')
-    def test_find_mac(self):
+    @unittest.skipIf(_AIX, 'AIX has different DELIM')
```
Rough inverse of the above suggestion:

    @unittest.skipUnless(self.uuid._MAC_DELIM == b':')

ncoghlan commented on Lib/uuid.py:
```diff
@@ -58,6 +58,7 @@
 _DARWIN  = platform.system() == 'Darwin'
 _LINUX   = platform.system() == 'Linux'
 _WINDOWS = platform.system() == 'Windows'
+_MAC_DELIM =  b':' if not _AIX else b'.'
```
Add:

    _MAC_OMITS_LEADING_ZEROES = _AIX

taleinat replied:
This is unnecessary with the current implementation.

ncoghlan commented on Lib/uuid.py:
```diff
@@ -412,20 +420,71 @@ def _find_mac(command, args, hw_identifiers, get_index):
+                    # en0   1500  link#2      fa.bc.de.f7.62.4 110854824     0 160133733     0     0
+                    # not
+                    # en0   1500  link#2      fa.bc.de.f7.62.04 110854824     0 160133733     0     0
+                    elif _AIX and  value.count(_MAC_DELIM) == 5:
```
And this can change to `_MAC_OMITS_LEADING_ZEROES` rather than being a direct reference to `_AIX`

ncoghlan commented on Lib/uuid.py:
```diff
@@ -412,20 +420,71 @@ def _find_mac(command, args, hw_identifiers, get_index):
+                    value = values[f_index(i)]
+                    if len(value) == 17 and value.count(_MAC_DELIM) == 5:
+                        mac = int(value.replace(_MAC_DELIM, b''), 16)
+                    # (Only) on AIX the macaddr value given is not prefixed by 0, e.g.
```
With the revised logic, `_AIX` is just the only currently known example of this behaviour, so the "(Only)" qualifier won't be needed in the comment.

taleinat replied:
I agree, but @vstinner explicitly said that this be made AIX-specific, though I imagine that he would agree with this suggestion. @vstinner?

ncoghlan commented on Misc/NEWS.d/next/Library/2018-08-04-12-26-11.bpo-28009.4JcHZb.rst:
```diff
@@ -0,0 +1 @@
+Fix uuid.getnode() on AIX - to use proper MAC_DELIM character
```
```suggestion
Fix uuid.getnode() on AIX to use proper MAC_DELIM character
and to handle the potential lack of leading zeroes in MAC address
segments.
```

aixtools replied:
I forget what was supposed to be there.  This works too.

ncoghlan reviewed (changes requested):
Thanks @aixtools, this is looking pretty nice now. Just a few structural requests aimed at reducing the likely number of "What's special about AIX here?" questions from future maintainers by defining the characteristics of interest where AIX differs, and then focusing the code and test logic on those, rather than checking for "Is this AIX?" directly.

aixtools committed:
Improve blurb

taleinat committed:
refactor more of getting a command's stdout into the helper func

taleinat committed:
reduce scope of try/except blocks, for clarity

taleinat committed:
make tests run regardless of host OS

taleinat committed:
remove no longer necessary _AIX definition in the test file

taleinat committed:
fix unittest import and indentation in one place

taleinat committed:
simplify macaddr parsing and control flow in _find_mac_nextlines()

This brings it more in line with _find_mac_inline().

taleinat committed:
no need for extra return value check in _netstart_getnode()

taleinat committed:
improve wording of NEWS entry

taleinat reviewed (approved):
With my recent changes, this LGTM.

I'd like to get approval from another core dev, though. Or perhaps @zestyping (the original author of the uuid module)?

taleinat committed:
make AIX formatted MAC address handling AIX-specific

This as per Victor Stinner's explicit request.

zestyping commented on Lib/uuid.py:
```diff
@@ -384,48 +394,103 @@ def _popen(command, *args):
 
-def _find_mac(command, args, hw_identifiers, get_index):
+
+# In the next two fucnctions:
```
I think I understand the general approach here, and it makes sense.  These two functions need a better explanation of what they do, though—the comments here aren't enough for a reader to understand how they're intended to work, and the argument names `hw_identifiers` and `f_index` are quite cryptic.

Here's an attempt at some better names and a docstring:

    def _find_mac_near_keyword(command, args, keywords, get_word_index):
        """Searches the output of a command for a MAC address near a keyword.

        Each line of words in the output is lowercased and searched for any of
        the given keywords.  Upon a match, get_word_index is invoked to pick a
        word from the line, given the index of the match (e.g. lambda i: 0 gets the
        first word on the line; lambda i: i - 1 gets the word preceding the keyword)."""

Given that `_find_mac_nextlines` is supposed to look for words under column headings, there's no need for it to take an `f_index` function, and the third argument can be named more clearly.  Indeed, it would be misleading to call the third argument `hw_identifiers`, as it's a single string, not a list of strings.  Here's a try:

    def _find_mac_under_heading(command, args, heading):
        """Looks for a MAC address under a heading in the output of a command.

        The first line of words in the output is searched for the given heading.
        Words at the same word index as the heading in subsequent lines are then
        examined to see if they look like MAC addresses."""
        
Another concern I have is that the logic about what looks like a MAC address is mixed in with the logic about where to find it.  It would be ideal to factor it out, like this:

    def _find_near_keyword(command, args, keywords, get_word_index):
        """Yields selected nearby words for keywords found in the output of a command.
        ...
            yield _parse_mac(word)
        ...

    def _find_mac_under_heading(command, args, heading):
        """Yields words found under a heading in the output of a command.

        The first line of words in the output is searched for the given heading.
        For each subsequent line, the word at the same word index as the heading
        in subsequent lines is then yielded."""
        
    def _parse_mac(word):
        """Converts a string to a 48-bit integer if it looks like a MAC address, or returns None."""
        # A MAC address is six bytes, each printed as one or two hex digits.
        # They may be separated by ':' or '.' depending on the platform.
        parts = word.replace('.', ':').split(':')
        if len(parts) == 6 and all(len(part) <= 2 for part in parts):
            hexstr = b''.join(part.rjust(2, b'0') for part in parts)
            try:
                return int(hexstr, 16)
            except ValueError:
                pass

    def _select_best_mac(macs):
        """Chooses a MAC from the given iterable, preferring any universally administered
        MAC address over locally administered MAC addresses."""
        local_macs = []
        for mac in macs:
            if _is_universal(mac):
                return mac
            elif mac:
                local_macs.append(mac)
        if local_macs:
            return local_macs[0]

Then:

    _select_best_mac(_find_near_keyword('ifconfig', args, keywords, lambda i: i + 1))
    _select_best_mac(_find_under_heading('netstat', '-ia', b'Address')

etc.

aixtools replied:
Thanks. 
Very interesting to read up better ways to do things, e.g., "parts = word.replace('.', ':').split(':')"

aixtools committed:
Stop reverse DNS lookups with netstat

zestyping commented on Lib/test/test_uuid.py:
```diff
@@ -674,27 +673,60 @@ class TestUUIDWithExtModule(BaseTestUUID, unittest.TestCase):
+
+        # The above data is specific to AIX - with '.' as _MAC_DELIM
+        # and strings shorter than 17 bytes (no leading 0).
+        # The above data will only be parsed properly on non-AIX unixes.
```
I'm puzzled why this comment says `non-AIX`—if the above data is an AIX example, is this meant to say `parsed properly on AIX` instead of `non-AIX`?

zestyping commented on Lib/uuid.py:
```diff
@@ -347,24 +348,33 @@ def version(self):
+    try:
+        executable = shutil.which(command)
         if executable is None:
+            path = os.pathsep.join(('/sbin', '/usr/sbin'))
```
You could use the built-in capability of `which` to search extra paths here:

    dirs = os.environ.get('PATH', os.defpath).split(os.pathsep)
    extended_path = os.pathsep.join(dirs + ['/sbin', '/usr/sbin'])
    executable = shutil.which(command, path=extended_path)

zestyping reviewed (commented):
Hi!  Here are my two cents.  I realize my suggestions involve substantial changes and I don't have prior context on the evolution of this PR, so feel free to take with a grain of salt.  I hope they are somewhat helpful, though.  I don't have authority over this file, merely opinions.

As a general note, it seems to me that it's more robust to write code that works for all cases (e.g. split on both colons _and_ periods; always tolerate hex bytes that have one _or_ two characters) rather than different behaviours for each platform, which might be more likely to break if the platform detection is incorrect or if the format changes slightly on a given platform.

taleinat committed:
add _MAC_OMITS_LEADING_ZEROES, better function names and doc-strings

aixtools committed:
update blurb and comments in test_uuid.py

aixtools committed:
Update News blurb

ncoghlan reviewed (approved):
There are still a few test skips that I think would be better off checking for the specific assumption(s) they're making, rather than checking for AIX as a platform, but I don't think those are a merge blocking issue.
